src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -3367,14 +3367,46 @@ public override Guid GUID
         {
             get
             {
+                TypeHandle th = GetNativeTypeHandle();
+                if (th.IsTypeDesc || IsArray)
+                {
+                    return Guid.Empty;
+                }
+
                 Guid result = default;
-                GetGUID(ref result);
+                unsafe
+                {
+#if FEATURE_COMINTEROP
+                    if (this == (RuntimeType)typeof(__ComObject))
+                    {
+                        GetComObjectGuidWorker(this, &result);
```

Comment:
This means that this effectively returns `GetComObjectGuidWorker(typeof(__ComObject), &result);`. It is going to be the same Guid for any COM object. Is that right? Looks like a bug or pre-existing bug.

Reply:
> It is going to be the same Guid for any COM object. Is that right?

Note in this case, that is why I need to pass in the object instance as well. For `__ComObject` types, the `SyncBlock` on the object instance is queried for the associated GUID - see `ReflectionInvocation_GetComObjectGuid`. That is why I separated this path from the other, so I could have a "fast" `MethodTable*` path and a "slow" object instance path.

Reply:
Do we have any test coverage for this?

I am not following how this works. `this == (RuntimeType)typeof(__ComObject)` should return true only if `this` is `typeof(__ComObject)` instance. I do not expect it to return true for the COM-type specific `RuntimeType` instances allocated by `AllocateComClassObject`.

Reply:
> Do we have any test coverage for this?

Yes, I did have failures on this code path. I will take another look tomorrow and verify my understanding here.

Reply:
I think the problem is that `IsComObjectClass(type)` check in the existing code is not equivalent to `this == (RuntimeType)typeof(__ComObject)` check.

Reply:
Isn't `return pMT == g_pBaseCOMObject;` that check?

Reply:
I see what you are getting at. I think I'm missing some subtle nuance with these "Is" functions/properties.

Reply:
There is 1:N relationship between `__ComObject`  MethodTable and RuntimeType instances.

Condition that covers all RuntimeType instances that point to `__ComObject` methodtable would have to be written as `th.AsMethodTable() == MethodTable.Of<__ComObject>()`.

Also, I am not sure about the purpose of this condition: https://github.com/dotnet/runtime/blob/3ca4ec8011f153bb8aa7508ea323a9acfe6e8283/src/coreclr/vm/interoputil.cpp#L797 . It is either unnecessary and the implementation can be simplified to just ``return type == TypeHandle(g_pBaseCOMObject);`, or there is some nuance that I am missing.